Improve JSON encoding

rcoup opened the pull request:
1. use the postgres builtin `escape_json()` function
2. escape all type, schema, table, field names. Add another test. Fixes #35 
3. escape generic logical decoding messages
4. leave `\x` alone unless it's a `bytea` prefix. Fixes #23, though like the reporter I'm not entirely convinced wal2json should be dropping the `\x` at all?

There's still a lot of duplicated code paths between "pretty" and "normal" modes. Personally I'd suggest dropping pretty-print mode from wal2json altogether and suggesting users pass it through [jq](https://stedolan.github.io/jq/) for debugging:
```
$ pg_recvlogical --slot test --start -o -o write-in-chunks=0 -f - | jq
```

Otherwise maybe some helper macros/functions that take care of the `\n\t` stuff so the duplication can be eliminated?

eulerto commented:
@rcoup Great! Let me review your PR.

* I'm not sure we want special test for escaping. How about using one of the existent tests (like insert1)?
* Use appendStringInfoChar() when adding commas instead of appendStringInfoString();
* Prefer contentbuf or content_buf instead of contentBuf (just to be consistent);
* Do we need a null-terminated content in strncpy()? I don't think so.

I'll appreciate if you could break this PR into 3 patches:
1) escape_json() + #35
2) #23 
3) escape generic WAL messages

It is easier to review and apply.

As for compact x pretty-print, I agree that compact format is the prefer mode (that is why I implemented it in commit f9446d1 and it is also the default mode since then). However, pretty print mode is useful for human inspections (I can't use `jq` in SQL interface). Avoid code duplication is always a goal to be achieved. Let's discuss the possibilities in another issue.

rcoup commented:
> Do we need a null-terminated content in strncpy()? I don't think so.

AFAICT there's no guarantee `content` is a null-terminated string, and `escape_json()` only deals with null-terminated strings.

> Use appendStringInfoChar() when adding commas instead of appendStringInfoString();

Hmm, I noticed `escape_json()` actually uses [`appendStringInfoCharMacro()`](https://doxygen.postgresql.org/stringinfo_8h_source.html#l00127), worth swapping to that?

eulerto commented:
@rcoup `content` does not contain the last character from the string. You can use one or another (indeed `appendStringInfoCharMacro` could use `appendStringInfoChar`). That code path is not used a lot that is why I suggested `appendStringInfoChar`.

rcoup commented:
Some updates:

* `pg_logical_emit_message()` actually takes a `bytea`, not text, and the `content` var is _not_ null-terminated. If we want to keep string outputs there is a conflict... since bytea is text-encoded as hex (`"\xC0FFEE"`). Currently I'm passing it through the json-encoder as a C-string, so special bytes <0x20 are escaped, but null bytes truncate, and the output isn't necessarily valid UTF-8.

  Pseudo-text output similar to `encode(E'\\x41203100C0FFEE'::bytea, 'escape') => 'A 1\000\300\377\356'` which would be a workaround doesn't _appear_ to be exposed in the PG C API (eg. `binary_encode()`), but maybe I'm not looking in the right place? Regardless, I've added tests with bytea messages so at least it's encoded what the current behaviour is.

  We could `\x` hex-encode if there's NULLs in the data, but that still leaves plenty of invalid UTF8 strings, and a consumer would randomly-sometimes-get hex.

* I've updated to use `appendStringInfoCharMacro()` on the fast-path, and `appendStringInfoChar()` everywhere else in the changes.

* And merged the new tests into the existing tests.

* Breaking up the patch is difficult if you want existing tests to pass at each stage – eg. #23 (`\x` change) is a side effect of swapping out `escape_json()`; I can't see a reason to keep no-escaping/overflow for generic wal messages when we fix other escaping uses; etc. I mean, I could keep `quote_escape_json()` and remove it in a later commit, but not sure it actually adds any value to the changeset.

rcoup commented:
@eulerto I've rebased against master again, tests pass :)

Any comments on the above issues?

rcoup commented:
@eulerto would love to see this land.

Pondering `pg_logical_emit_message()` again, doing the pseudo-text output similar to `encode(..., 'escape') => 'A 1\000\300\377\356'` wouldn't be too complex to do in wal2json.
